homeassistant/components/lock/__init__.py

```diff
@@ -72,18 +73,52 @@ async def async_setup(hass: HomeAssistant, config: ConfigType) -> bool:
     await component.async_setup(config)
 
     component.async_register_entity_service(
-        SERVICE_UNLOCK, LOCK_SERVICE_SCHEMA, "async_unlock"
+        SERVICE_UNLOCK, LOCK_SERVICE_SCHEMA, _async_unlock
     )
     component.async_register_entity_service(
-        SERVICE_LOCK, LOCK_SERVICE_SCHEMA, "async_lock"
+        SERVICE_LOCK, LOCK_SERVICE_SCHEMA, _async_lock
     )
     component.async_register_entity_service(
-        SERVICE_OPEN, LOCK_SERVICE_SCHEMA, "async_open"
+        SERVICE_OPEN, LOCK_SERVICE_SCHEMA, _async_open
     )
 
     return True
 
 
+async def _async_lock(entity: LockEntity, service_call: ServiceCall) -> None:
+    """Lock the lock."""
+    code: str = service_call.data.get(ATTR_CODE, "")
+    if entity.code_format_cmp and not entity.code_format_cmp.match(code):
+        raise ValueError(
+            f"Code '{code}' for locking {entity.name} doesn't match pattern {entity.code_format}"
+        )
+    await entity.async_lock(**service_call.data)
+
+
+async def _async_unlock(entity: LockEntity, service_call: ServiceCall) -> None:
+    """Unlock the lock."""
+    code: str = service_call.data.get(ATTR_CODE, "")
+    if entity.code_format_cmp and not entity.code_format_cmp.match(code):
+        raise ValueError(
+            f"Code '{code}' for unlocking {entity.name} doesn't match pattern {entity.code_format}"
+        )
+    await entity.async_unlock(**service_call.data)
+
+
+async def _async_open(entity: LockEntity, service_call: ServiceCall) -> None:
+    """Open the door latch."""
+    if not (entity.supported_features & LockEntityFeature.OPEN):
```

Comment:
This should not be checked here, it should be added to the service registration.

As an example, this is how it's done for `climate`:
```py
    component.async_register_entity_service(
        SERVICE_SET_PRESET_MODE,
        {vol.Required(ATTR_PRESET_MODE): cv.string},
        "async_set_preset_mode",
        [ClimateEntityFeature.PRESET_MODE],
    )
```

Reply:
Updated that, but for a group of locks I set this feature. When calling the service it will not be validated through the schema.
Should we not have both validations?